doc/instruction_decode_execute.rst

```diff
@@ -0,0 +1,111 @@
+
+ID
+-----------------------------
+Source File: :file:`rtl/ibex_id_stage.sv`
+
+The Instruction Decode (ID) controls the overall decode/execution process.
+It contains the muxes to choose what is sent to the ALU inputs and where the write data for the register file comes from.
+A small state machine is used to control multi-cycle instructions (see :ref:`pipeline-details` for more details), which stalls the whole stage whilst a multi-cycle instruction is executing.
+
+Controller
+----------
+Source File: :file:`rtl/ibex_controller.sv`
+
+The Controller contains the state machine that controls the overall execution of the processor.
+It is responsible for:
+
+* Handling core startup from reset
+* Setting the PC for the IF stage on jump/branch
+* Dealing with exceptions/interrupts (jump to appropriate PC, set relevant CSR values)
+* Controls sleep/wakeup on WFI
+* Debugging control
+
+Decoder
+-------
+Source File: :file:`rtl/ibex_decoder.sv`
+
+The decoder takes uncompressed instruction data and issues appropriate control signals to the other blocks to execute the instruction.
+
+Reg File
+--------
+Source Files: :file:`rtl/ibex_register_file_ff.sv` :file:`rtl/ibex_register_file_latch.sv`
+
+See :ref:`register-file` for more details.
+
+EX Block
+--------
+Source File: :file:`rtl/ibex_ex_block.sv`
+
+The EX block Contains the ALU and Mult/Div blocks, it does little beyond wiring and instantiating these blocks.
+
+ALU
+---
+Source File: :file:`rtl/ibex_alu.sv`
+
+The Arithmetic Logic Logic (ALU) is a purely combinational block that implements operations required for the Integer Computational Instructions and the comparison operations required for the Control Transfer Instructions in the RV32I RISC-V Specification.
+Other blocks use the ALU for the following tasks:
+
+* Mult/Div uses it to perform addition as part of the multiplication and division algorithms
+* It computes branch targets with a PC + Imm calculation
+* It computes memory addresses for loads and stores with a Reg + Imm calculation
+* The LSU uses it to increment addresses when performing two accesses to handle an unaligned access
+
+.. _mult-div:
+
+MULT/DIV
+--------
+Source Files: :file:`rtl/ibex_multdiv_slow.sv` :file:`rtl/ibex_multdiv_fast.sv`
+
+The Multiplier/Divider (MULT/DIV) is a state machine driven block to perform multiplication and division.
+The fast and slow versions differ in multiplier only, both implement the same form of long division algorithm.
+The ALU block is used by the long division algorithm in both the fast and slow blocks.
+
+Fast Multiplier
+  Completes multiply in 3-4 cycles using a MAC (multiply accumulate) which is capable of a 17-bit x 17-bit multiplication with a 34-bit accumulator.
+  MUL instruction takes 3 cycles, MULH takes 4.
+  This MAC is internal to the mult/div block (no external ALU use).
+  Beware it is simply implemented with the '*' and '+' operators so results heavily depend upon the synthesis tool used.
```

Comment:
`Beware that the implementation uses the ``*`` and ``*`` SystemVerilog operators, ...`